src/client/Microsoft.Identity.Client/AuthScheme/IAuthenticationOperation2.cs

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.AuthScheme
+{
+    /// <summary>
+    /// This is an extensibility API and should only be used by SDKs.
+    /// Enhanced version of IAuthenticationOperation that supports asynchronous token formatting.
+    /// Used to modify the experience depending on the type of token asked with async capabilities.
+    /// </summary>
+    public interface IAuthenticationOperation2 : IAuthenticationOperation
+    {
+        /// <summary>
+        /// Will be invoked instead of IAuthenticationOperation.FormatResult
+        /// </summary>
+        Task FormatResultAsync(AuthenticationResult authenticationResult, CancellationToken cancellationToken = default);
+    }
```

Comment:
Does it make sense to also make `IReadOnlyDictionary<string, string> GetTokenRequestParams()` async? As sometimes those parameters also need to be fetched from somewhere?

Reply:
Let's track this via https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/5361 instead, we need to look at this holistically.